lnwallet/btcwallet/signer.go

```diff
@@ -73,6 +76,94 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
 	return addr.(waddrmgr.ManagedPubKeyAddress).PrivKey()
 }
 
+// deriveKeyByBIP32Path derives a key described by a BIP32 path. The path is
+// expected to have 2^31 added to a hardened key in the path.
+func (b *BtcWallet) deriveKeyByBIP32Path(path []uint32) (*btcec.PrivateKey,
+	error) {
+
+	// Make sure we get a full path with exactly 5 elements. A path is
+	// either an internal one with 4 dynamic and one static elements:
+	//    m/1017'/coinType'/keyFamily'/0/index
+	// Or an external one with 5 elements:
+	//    m/purpose'/coinType'/account'/change/index
+	const minDerivationPathDepth = 5
+	if len(path) < minDerivationPathDepth {
+		return nil, fmt.Errorf("invalid BIP32 derivation path, "+
+			"expected minimum path length %d, instead was %d",
+			minDerivationPathDepth, len(path))
+	}
+
+	purpose := path[0] - hdkeychain.HardenedKeyStart
+	coinType := path[1] - hdkeychain.HardenedKeyStart
+	account := path[2] - hdkeychain.HardenedKeyStart
+	change, index := path[3], path[4]
+
+	// Is this an internal key?
```

Comment:
newbie-nit: the term "internal" makes me think of change addresses in this context (internal vs external)
ignore if we already use this terminology elsewhere, just a thought

Reply:
Hmm, good point. I rephrased "internal" and "external" to "custom purpose" and "default BIP" in this context to make it more clear. I hope I caught every occurrence.